cabot/cabotapp/models.py

```diff
@@ -1146,6 +1169,146 @@ def __unicode__(self):
+
+    @classmethod
+    def get_acks_matching_check(cls, check, at_time=None):
+        # type: (StatusCheck, timezone.datetime) -> List[Acknowledgement]
+        """
+        :param check: check to gather acks for
+        :param at_time: only consider acks that were open at this time; leave as None to use timezone.now().
+        :returns list of Acknowledgements where ack.
+        """
+        at_time = at_time or timezone.now()
+        acks = cls.objects.filter(status_check_id=check.id)
+        return acks.exclude(created_at__gt=at_time).exclude(closed_at__lte=at_time).exclude(expire_at__lte=at_time)
+
+    @classmethod
+    def get_acks_matching_result(cls, result, at_time=None):
+        # type: (StatusCheckResult, Union[timezone.datetime, None]) -> List[Acknowledgement]
+        """
+        :param result: result to gather acks for
+        :param at_time: only consider acks that were open at this time; leave None for all currently open acks
+        :returns list of Acknowledgements where ack.matches_result(result) == True
+        """
+        return [a for a in cls.get_acks_matching_check(result.status_check, at_time) if a.matches_result(result)]
+
+    @classmethod
+    def close_succeeding_acks(cls, check, at_time=None):
+        # type: (StatusCheck, Union[timezone.datetime, None]) -> None
+        """
+        This function closes acks that have hit their close_after_successes threshold.
+        This should be called after a StatusCheck saves a new StatusCheckResult (including its tags!).
+        It technically only needs to be called after a check succeeds, but it's idempotent - you can call it whenever.
+        :param check: StatusCheck to update acks for.
+        :param at_time: Time to pass into get_acks_matching_check. Leave as None to use timezone.now().
+        """
+        for ack in cls.get_acks_matching_check(check, at_time=at_time):
+            if ack.close_after_successes:
+                # get last n results for our check, regardless of status
+                results = StatusCheckResult.objects.filter(status_check=ack.status_check)\
+                              .order_by('-time_complete', '-id').only('succeeded')[:ack.close_after_successes]
+
+                # now filter down to the results that succeeded
+                results = [r for r in results if r.succeeded]
+
+                # if we hit our threshold, this check should expire
+                if len(results) >= ack.close_after_successes:
+                    reason = 'check passed {} times'.format(len(results)) if len(results) != 1 else 'check passed'
+                    ack.close(reason)
+
+    def save(self, **kwargs):
+        # THERE CAN BE ONLY ONE. for log trails.
+        existing = Acknowledgement.objects\
+            .filter(status_check=self.status_check, closed_at__isnull=True)\
+            .exclude(pk=self.pk)
+        for old_ack in existing:
+            old_ack.close('options changed')
```

Comment:
When someone edits an ack in the UI, it actually creates a new object and the old one is closed here. This keeps the options on acks immutable as a lazy way to keep a history of what was acked when, by who, and with what options.

An event table would be nicer, but I feel takes a bit more work to implement the logic to answer queries like "why was this check considered acked at time X" (which I suspect is the most common log checking we'd do). Since ack creation should be pretty low-traffic and the model is pretty small, I thought just creating new rows would be okay.

Reply:
ok, still think the "options changed" wording is kind of confusing--maybe "ack updated" (ack recreated)?